python/cudf/cudf/_lib/ast.pyx

```diff
@@ -0,0 +1,348 @@
+    SIN = libcudf_ast.ast_operator.SIN
+    COS = libcudf_ast.ast_operator.COS
+    TAN = libcudf_ast.ast_operator.TAN
+    ARCSIN = libcudf_ast.ast_operator.ARCSIN
+    ARCCOS = libcudf_ast.ast_operator.ARCCOS
+    ARCTAN = libcudf_ast.ast_operator.ARCTAN
+    SINH = libcudf_ast.ast_operator.SINH
+    COSH = libcudf_ast.ast_operator.COSH
+    TANH = libcudf_ast.ast_operator.TANH
+    ARCSINH = libcudf_ast.ast_operator.ARCSINH
+    ARCCOSH = libcudf_ast.ast_operator.ARCCOSH
+    ARCTANH = libcudf_ast.ast_operator.ARCTANH
+    EXP = libcudf_ast.ast_operator.EXP
+    LOG = libcudf_ast.ast_operator.LOG
+    SQRT = libcudf_ast.ast_operator.SQRT
+    CBRT = libcudf_ast.ast_operator.CBRT
+    CEIL = libcudf_ast.ast_operator.CEIL
+    FLOOR = libcudf_ast.ast_operator.FLOOR
+    ABS = libcudf_ast.ast_operator.ABS
+    RINT = libcudf_ast.ast_operator.RINT
+    BIT_INVERT = libcudf_ast.ast_operator.BIT_INVERT
+    NOT = libcudf_ast.ast_operator.NOT
+
+
+class TableReference(Enum):
+    LEFT = libcudf_ast.table_reference.LEFT
+    RIGHT = libcudf_ast.table_reference.RIGHT
+
+
+# Note that this function only currently supports numeric literals. libcudf
+# expressions don't really support other types yet though, so this isn't
+# restrictive at the moment.
+cdef class Literal(Expression):
+    def __cinit__(self, value):
+        # TODO: Would love to find a better solution than unions for literals.
+        cdef int intval
+        cdef float floatval
+
+        if isinstance(value, int):
+            self.c_scalar_type = scalar_type_t.INT
+            intval = value
+            self.c_scalar.int_ptr = make_unique[numeric_scalar[int64_t]](
+                intval, True
+            )
+            self.c_obj = <expression_ptr> make_unique[libcudf_ast.literal](
+                <numeric_scalar[int64_t] &>dereference(self.c_scalar.int_ptr)
+            )
+        elif isinstance(value, float):
+            self.c_scalar_type = scalar_type_t.DOUBLE
+            floatval = value
+            self.c_scalar.double_ptr = make_unique[numeric_scalar[double]](
+                floatval, True
+            )
+            self.c_obj = <expression_ptr> make_unique[libcudf_ast.literal](
+                <numeric_scalar[double] &>dereference(self.c_scalar.double_ptr)
+            )
+
+
+cdef class ColumnReference(Expression):
+    def __cinit__(self, size_type index):
+        self.c_obj = <expression_ptr>make_unique[libcudf_ast.column_reference](
+            index
+        )
+
+
+cdef class Operation(Expression):
+    def __cinit__(self, op, Expression left, Expression right=None):
+        # This awkward double casting is the only way to get Cython to generate
+        # valid C++. Cython doesn't support scoped enumerations, so it assumes
+        # that enums correspond to their underlying value types and will thus
+        # attempt operations that are invalid without first explicitly casting
+        # to the underlying before casting to the desired type.
```

Comment:
This is an issue throughout the codebase where maybe it makes sense to document in a more central place.

Reply:
Yeah, I think when I wrote this code the first time I had to inspect the generated C++ manually to understand what was happening. It wasn't until later that I worked on other parts of the code using enums that had the same issue and I saw that we did the same casting elsewhere. Where do you think would be a good place for this?

Reply:
Yeah I guess there isn't a great place I can think of other than a Cython section in the devguide. We can leave this here, it's no big deal.

Reply:
I'll leave it for now. We can revisit as part of writing the dev guide. But also hopefully we can make Cython 3.0 happen soon enough to switch! Then we could use scoped enums, which fix this I believe.